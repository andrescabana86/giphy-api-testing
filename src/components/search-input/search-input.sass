.search-input
  display: block
  margin: 0 4px
  padding: 0
  border: 0
  outline: none