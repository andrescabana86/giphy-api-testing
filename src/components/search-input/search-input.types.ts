export type TSearchInputProps = {
  onSearch(value: string): void
  onClear(): void
}
