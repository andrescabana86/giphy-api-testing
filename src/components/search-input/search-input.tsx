import React, { ChangeEvent, useCallback, useEffect, useState, KeyboardEvent } from 'react'
import { TSearchInputProps } from './search-input.types'
import './search-input.sass'

export const SearchInput = React.memo((props: TSearchInputProps) => {
  const { onSearch, onClear } = props
  const [value, setValue] = useState('')
  const [lastKeyPessed, setLastKeyPressed] = useState<string | undefined>()

  const onChange = useCallback((evt: ChangeEvent<HTMLInputElement>) => {
    const value = evt.target.value
    setValue(value)
  }, [])

  const onKeyDown = useCallback((evt: KeyboardEvent<HTMLInputElement>) => {
    const lastKeyPressed = evt.key
    setLastKeyPressed(lastKeyPressed)
  }, [])

  useEffect(() => {
    const handler = setTimeout(() => {
      if (value.length > 2) {
        onSearch(value)
      } else if (!value.length && lastKeyPessed === 'Backspace') {
        onClear()
      }
    }, 500)

    return () => {
      clearTimeout(handler)
    }
  }, [value, lastKeyPessed])

  return (
    <input
      className="search-input"
      data-testid="search-input"
      onChange={onChange}
      type="text"
      value={value}
      onKeyDown={onKeyDown}
    />
  )
})
