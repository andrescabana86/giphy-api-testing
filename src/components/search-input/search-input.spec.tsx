import React from 'react'
import { render, cleanup, fireEvent } from '@testing-library/react'
import { SearchInput } from './search-input'

jest.useFakeTimers()

describe('SearchInput', () => {
  afterEach(cleanup)

  it('should render SearchInput component', () => {
    // Arrange
    const onSearch = () => null
    const onClear = () => null

    // Act
    const { asFragment } = render(<SearchInput onSearch={onSearch} onClear={onClear} />)

    // Assert
    expect(asFragment()).toMatchSnapshot()
  })

  it('should call onSearch when user type something and after 500ms', () => {
    // Arrange
    const onSearch = jest.fn(() => {
      console.log('call')
    })
    const onClear = () => null
    const { getByTestId } = render(<SearchInput onSearch={onSearch} onClear={onClear} />)
    const input = getByTestId('search-input')

    // Act
    input.onchange
    fireEvent.change(input, { target: { value: 'playstation' } })

    // Assert
    expect(onSearch).toHaveBeenCalledTimes(0)
    jest.runAllTimers()
    expect(onSearch).toHaveBeenCalledTimes(1) // func not called
  })
})
