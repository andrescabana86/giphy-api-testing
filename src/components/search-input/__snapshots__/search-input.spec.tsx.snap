// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchInput should render SearchInput component 1`] = `
<DocumentFragment>
  <input
    class="search-input"
    data-testid="search-input"
    type="text"
    value=""
  />
</DocumentFragment>
`;
